File: runtime/python/units/wasi/path_remove_directory.py
# requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
def wasi_path_remove_directory(self, dirfd, path_ptr, path_len):
    rel = self.memory.read_string(path_ptr, path_len).decode("utf-8", "surrogateescape")
    # rmdir(2) never follows a trailing symlink.
    host_path, err = self.resolve_path(dirfd, rel, False)
    if err is not None:
        return err
    # rmdir through a trailing slash on an existing directory is EINVAL per
    # wasmtime; other shapes come from the host call.
    if host_path.endswith(os.sep) and os.path.isdir(host_path[:-1]):
        return self.ERRNO_INVAL
    try:
        os.rmdir(host_path)
    except OSError as e:
        return self.fs_errno(e)
    return self.ERRNO_SUCCESS

File: runtime/go/units/wasi/path_remove_directory.go
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
func (w *WASI) wasi_path_remove_directory(dirfd, pathPtr, pathLen uint32) uint32 {
    rel := string(w.memory.read_string(uint64(pathPtr), uint64(pathLen)))
    // rmdir(2) never follows a trailing symlink; use the raw syscall so it
    // fails (ENOTDIR) on a non-directory rather than unlinking it.
    hostPath, err := w.resolve_path(dirfd, rel, false)
    if err != wasiOk {
        return err
    }
    // rmdir through a trailing slash on an existing directory is EINVAL per
    // wasmtime; other shapes come from resolve_path or the syscall.
    if strings.HasSuffix(rel, "/") {
        if fi, e := os.Stat(hostPath); e == nil && fi.IsDir() {
            return wasiInval
        }
    }
    if e := syscall.Rmdir(hostPath); e != nil {
        return w.fs_errno(e)
    }
    return wasiOk
}
